src/connector/src/sink/pulsar.rs

```diff
@@ -0,0 +1,321 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{HashMap, VecDeque};
+use std::fmt::Debug;
+use std::time::Duration;
+
+use anyhow::anyhow;
+use futures::future::try_join_all;
+use futures::TryFutureExt;
+use serde::Deserialize;
+use async_trait::async_trait;
+use futures_async_stream::for_await;
+use pulsar::{Pulsar, TokioExecutor, Producer};
+use pulsar::producer::{SendFuture, Message};
+use risingwave_common::catalog::Schema;
+use risingwave_common::array::StreamChunk;
+use risingwave_rpc_client::ConnectorClient;
+
+use super::{
+    Sink, SinkError, SinkWriterParam, SinkWriter, SinkParam, 
+    SINK_TYPE_OPTION, SINK_TYPE_APPEND_ONLY, SINK_TYPE_UPSERT,
+};
+use super::encoder::{JsonEncoder, TimestampHandlingMode};
+use crate::{deserialize_u32_from_string, deserialize_duration_from_string};
+use crate::sink::{Result, DummySinkCommitCoordinator};
+use crate::sink::utils::{
+    gen_append_only_message_stream, gen_upsert_message_stream,
+    AppendOnlyAdapterOpts, UpsertAdapterOpts,
+};
+use crate::common::PulsarCommon;
+
+pub const PULSAR_SINK: &str = "pulsar";
+
+/// The delivery buffer queue size
+/// When the `SendFuture` the current `send_future_buffer`
+/// is buffering is greater than this size, then enforcing commit once
+const PULSAR_SEND_FUTURE_BUFFER_MAX_SIZE: usize = 65536;
+
+const fn _default_max_retries() -> u32 {
+    3
+}
+
+const fn _default_retry_backoff() -> Duration {
+    Duration::from_millis(100)
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct PulsarConfig {
+    #[serde(
+        rename = "properties.retry.max",
+        default = "_default_max_retries",
+        deserialize_with = "deserialize_u32_from_string"
+    )]
+    pub max_retry_num: u32,
+
+    #[serde(
+        rename = "properties.retry.interval",
+        default = "_default_retry_backoff",
+        deserialize_with = "deserialize_duration_from_string"
```

Comment:
may try with `serde_as`, I dont know if it works with `Duration`

Reply:
Unable to use serve_as here because Duration does not implement FromStr trait.